Add JSX Support (#59)
alden12/rxfm: README.md

# RxFM - A Web Framework Built on RxJS

![Node.js CI](https://github.com/alden12/rxfm/workflows/Node.js%20CI/badge.svg?branch=master)
[![NPM](https://img.shields.io/npm/v/rxfm)](https://www.npmjs.com/package/rxfm)
[![Bundlephobia](https://img.shields.io/bundlephobia/minzip/rxfm?label=gzipped)](https://bundlephobia.com/result?p=rxfm@latest)
[![MIT license](https://img.shields.io/npm/l/rxfm)](https://opensource.org/licenses/MIT)

> RxFM *(working title)* is an experimental web framework born out of a wish for better [RxJS](https://github.com/ReactiveX/rxjs) integration, greater simplicity, and improved transparency in what a framework is doing under the hood.

Author: ```suggestion
RxFM is an experimental web framework born out of a wish for better [RxJS](https://github.com/ReactiveX/rxjs) integration, greater simplicity, and improved transparency in what a framework is doing under the hood.
```

```typescript
HelloWorld.subscribe(el => document.body.appendChild(el));
```

The root component should be the only subscribed component in our application, and indeed ideally the only use of `subscribe` at all! All being well, other observables should piggyback on the application subscription and are subscribed by virtue of being a part of the component stream. This way a single subscription at the app root can set the entire application in motion!

### Component Children:

We can pass lots of different kinds of things as component children:

```jsx
const ChildrenExample = () => <div>
  Children can be strings, 
  <b> child components, </b>
  or observables: {timer(0, 1000)}s elapsed.
</div>;
```

We can write code inside our JSX tags by using curly braces.

### Custom JSX Elements

In order to define components which can be used as JSX elements, they simply need to be declared as functions returning other JSX elements. For example we can use the `ChildrenExample` above as a JSX element. Elements with no children don't need to provide a closing tag.

```jsx
const CustomJsxElementExample = () => <ChildrenExample />;
```

> Data can be passed into a JSX element using props, these props are then available as an object in the first argument of a component function. Component children can be accessed using the `children` prop which is implicitly available to all JSX elements. To use props with TypeScript, a component function can be assigned the `FC` (function component) type imported from RxFM. The interface passed to `FC` will define the props available on a component as shown below.

Author: ```suggestion
Data can be passed into a JSX element using props, these props are then available as an object in the first argument of a component function. Component children can be accessed using the `children` prop which is implicitly available to all JSX elements. To use props with TypeScript, a component function can be assigned the `FC` (function component) type imported from RxFM. The generic type passed to `FC` will define the props available on a component as shown below.
```

```tsx
import RxFM, { FC } from 'rxfm';

> const FunFacts: FC<{ name: string }> = ({ name, children }) => <div>

Author: ```suggestion
interface FunFactsProps {
  name: string;
}

const FunFacts: FC<FunFactsProps > = ({ name, children }) => <div>
```

```

With `switchMap` we can map an observable to a component observable depending on a condition, or to an observable emitting one of either: `null | undefined | false` to remove it from the DOM.

```jsx
import { switchMap } from 'rxjs/operators';
import { of } from 'rxjs';

const ConditionalComponentsExample = () => <div>
  {flipFlop.pipe(
    switchMap(visible => visible ? <div>Now you see me!</div> : of(null)),
  )}
</div>;
```

> This may also be written more simply as: `conditional(flipFlop, <div>Now you see me!</div>)` using the `conditional` helper function from rxfm.

Author: ```suggestion
This may also be written more simply as below using the `conditional` helper function from rxfm:

```jsx
conditional(flipFlop, <div>Now you see me!</div>)
```
```

You may also be tempted to use `switchMap` to transform and array observable into an array of components (similar to using Array.map in React), but this will be rather inefficient as the components will be recreated each time the observable emits. The `mapToComponents` operator function should be used instead in this case as this will ensure that components are only recreated when necessary (see the [Dynamic Component Arrays](#dynamic-component-arrays) section).

[Code](https://github.com/alden12/rxfm/blob/master/src/app/basic-examples/conditional-components.ts) | [Live Demo](https://alden12.github.io/rxfm/)

---

## Component Outputs

We've seen how to pass props to a component in the [Custom JSX Elements](#custom-jsx-elements) section. Outputs can be provided by passing in callback functions as props to handle events inside the component.

```typescript
interface OptionButtonProps {
  option: string;
  setOption: (option: string) => void;
  active: Observable<boolean>;
}
```

> ```tsx

Author: ```suggestion
Here each `OptionButton` component takes a `setOption` callback prop to set itself as the active option.
```tsx
```

const OptionButton: FC<OptionButtonProps> = ({ option, setOption, active }) => {

>   const handleClick = () => setOption(option);
> 
>   const activeClassName = active.pipe(
>     map(active => active && 'active')
>   );
> 
>   return <button onClick={handleClick} class={['option-button', activeClassName]}>

Author: ```suggestion
  const activeClassName = active.pipe(
    map(active => active && 'active')
  );

  return <button onClick={() => setOption(option)} class={['option-button', activeClassName]}>
```

    {option}
  </button>;
};
```

> ```tsx

Author: ```suggestion
We can store then `selectedOption` state in the component wrapping the option buttons and pass in the `setOption` callback to each `OptionButton` instance.
```tsx
```

const options = ['Option 1', 'Option 2', 'Option 3'];

const ComponentOutputsExample = () => {
  const selectedOption = new BehaviorSubject<string>('Option 1');
  const setOption = (option: string) => selectedOption.next(option);

  const optionButtons = options.map(option => {
    const active = selectedOption.pipe(map(selectedOpt => selectedOpt === option));

>     return <OptionButton option={option} setOption={setOption} active={active} />;

Author: ```suggestion
    return <OptionButton option={option => selectedOption.next(option)} setOption={setOption} active={active} />;
```

]);
```

We can then define a component taking an individual item observable:

```tsx
const Item: FC<{ item: Observable<TodoItem> }> = ({ item }) => <div>
  {item.pipe(
    map(({ name, done }) => `${name} is ${done ? '' : 'not'} done!`),
  )}
</div>;
```

Using the `mapToComponents` operator function, we can map the item array into an array of `Item` components. Here the second argument is either a function taking the item and returning its unique id (similar to the 'key' prop in React) or a property name on the item where its unique id can be found. If this argument is omitted then the item's index in the source array will be used as its id.

```tsx
const ItemComponents = items.pipe(
    mapToComponents(item => <Item item={item} />, 'name'),
  );
```

The resulting component array observable can be passed directly as a component child:
```typescript
const ComponentArraysExample = <div>{ItemComponents}</div>;
```

> If our `items` subject were to then emit a new array, this would be immediately be reflected by our `Item` components in the DOM. Any items with matching ids from the previous emission will reuse the existing DOM elements. 

Author: ```suggestion
If our `items` subject were to then emit a new array, this would immediately be reflected by our `Item` components in the DOM. Any items with matching ids from the previous emission will reuse the existing DOM elements. 
```
